File: victorious/victorious/Categories/NSDate+timeSince.m
//
//  NSDate+timeSince.m
//  victorious
//
//  Created by Gary Philipp on 1/9/14.
//  Copyright (c) 2014 Victorious. All rights reserved.
//

#import "NSDate+timeSince.h"

@implementation NSDate (timeSince)

- (NSString *)stringDescribingTimeIntervalSinceNow
{
    return [self stringDescribingTimeIntervalSinceNowWithPrecision:VTimeSincePrecisionMinutes];
}

- (NSString *)stringDescribingTimeIntervalSince:(NSDate *)date
{
    return [self stringDescribingTimeIntervalSince:date precision:VTimeSincePrecisionMinutes];
}

- (NSString *)stringDescribingTimeIntervalSinceNowWithPrecision:(VTimeSincePrecision)precision
{
    return [self stringDescribingTimeIntervalSince:[NSDate date] precision:precision];
}

- (NSString *)stringDescribingTimeIntervalSince:(NSDate *)date precision:(VTimeSincePrecision)precision
{
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *components = [calendar components:
                                    NSCalendarUnitYear|
                                    NSCalendarUnitMonth|
                                    NSCalendarUnitWeekOfMonth|
                                    NSCalendarUnitDay|
                                    NSCalendarUnitHour|
                                    NSCalendarUnitMinute|
                                    NSCalendarUnitSecond
                                               fromDate:self
                                                 toDate:date
                                                options:0];
    
    if (components.year >= 1)
    {
        if (components.year == 1)
        {
            return NSLocalizedString(@"LastYear", @"");
        }
        else
        {
            return [NSString stringWithFormat:NSLocalizedString(@"YearsAgo", @""), components.year];
        }
    }
    else if (components.month >= 1)
    {
        if (components.month == 1)
        {
            return NSLocalizedString(@"LastMonth", @"");
        }
        else
        {
            return [NSString stringWithFormat:NSLocalizedString(@"MonthsAgo", @""), components.month];
        }
    }
    else if (components.weekOfMonth >= 1)
    {
        if (components.weekOfMonth == 1)
        {
            return NSLocalizedString(@"LastWeek", @"");
        }
        else
        {
            return [NSString stringWithFormat:NSLocalizedString(@"WeeksAgo", @""), components.weekOfMonth];
        }
    }
    else if (components.day >= 1)    // up to 6 days ago
    {
        if (components.day == 1)
        {
            return NSLocalizedString(@"Yesterday", @"");
        }
        else
        {
            return [NSString stringWithFormat:NSLocalizedString(@"DaysAgo", @""), components.day];
        }
    }
    else if (components.hour >= 1)   // up to 23 hours ago
    {
        if (components.hour == 1)
        {
            return NSLocalizedString(@"HourAgo", @"");
        }
        else
        {
            return [NSString stringWithFormat:NSLocalizedString(@"HoursAgo", @""), components.hour];
        }
    }
    else if (components.minute >= 1) // up to 59 minutes ago
    {
        if (components.minute == 1)
        {
            return NSLocalizedString(@"MinuteAgo", @"");
        }
        else
        {
            return [NSString stringWithFormat:NSLocalizedString(@"MinutesAgo", @""), components.minute];
        }
    }
    else if (precision == VTimeSincePrecisionSeconds && components.second > 0)
    {
        if (components.second == 1)
        {
            return NSLocalizedString(@"SecondAgo", @"");
        }
        else
        {
            return [NSString stringWithFormat:NSLocalizedString(@"SecondsAgo", @""), components.second];
        }
    }
    else
    {
        return NSLocalizedString(@"Now", @"");
    }
}

@end

File: victorious/victorious/Categories/NSDate+timeSince.swift
//
//  NSDate+timeSince.swift
//  victorious
//
//  Created by Jarod Long on 4/7/16.
//  Copyright © 2016 Victorious. All rights reserved.
//

import Foundation

@objc enum DateTimeIntervalStringFormat: Int {
    case concise, verbose
}

@objc enum DateTimeIntervalStringPrecision: Int {
    case minutes, seconds
}

extension Date {
    // Avoiding default values for the format parameter for Objective-C compatibility.
    
    func stringDescribingTimeIntervalSinceNow() -> String {
        return stringDescribingTimeIntervalSinceNow(format: .concise, precision: .minutes)
    }
    
    func stringDescribingTimeIntervalSinceNow(format: DateTimeIntervalStringFormat, precision: DateTimeIntervalStringPrecision) -> String {
        return stringDescribingTimeIntervalSince(Date(), format: format, precision: precision)
    }
    
    func stringDescribingTimeIntervalSince(_ date: Date) -> String {
        return stringDescribingTimeIntervalSince(date, format: .concise, precision: .minutes)
    }
    
    func stringDescribingTimeIntervalSince(_ date: Date, format: DateTimeIntervalStringFormat, precision: DateTimeIntervalStringPrecision) -> String {
        let calendar = Calendar.current
        let components = (calendar as NSCalendar).components([.year, .month, .weekOfMonth, .day, .hour, .minute, .second], from: self, to: date, options: [])
        
        switch components {
        case let components where components.year! > 1:
            switch format {
            case .concise:
                return NSString(format: NSLocalizedString("YearsAgo", comment: "") as NSString, components.year!) as String
            case .verbose:
                return NSString(format: NSLocalizedString("YearsAgo Verbose", comment: "") as NSString, components.year!) as String
            }
        
        case let components where components.year == 1:
            switch format {
            case .concise:
                return NSLocalizedString("LastYear", comment: "")
            case .verbose:
                return NSLocalizedString("LastYear Verbose", comment: "")
            }
        
        case let components where components.month! > 1:
            switch format {
            case .concise:
                return NSString(format: NSLocalizedString("MonthsAgo", comment: "") as NSString, components.month!) as String
            case .verbose:
                return NSString(format: NSLocalizedString("MonthsAgo Verbose", comment: "") as NSString, components.month!) as String
            }
        
        case let components where components.month == 1:
            switch format {
            case .concise:
                return NSLocalizedString("LastMonth", comment: "")
            case .verbose:
                return NSLocalizedString("LastMonth Verbose", comment: "")
            }
        
        case let components where components.weekOfMonth! > 1:
            switch format {
            case .concise:
                return NSString(format: NSLocalizedString("WeeksAgo", comment: "") as NSString, components.weekOfMonth!) as String
            case .verbose:
                return NSString(format: NSLocalizedString("WeeksAgo Verbose", comment: "") as NSString, components.weekOfMonth!) as String
            }
        
        case let components where components.weekOfMonth == 1:
            switch format {
            case .concise:
                return NSLocalizedString("LastWeek", comment: "")
            case .verbose:
                return NSLocalizedString("LastWeek Verbose", comment: "")
            }
        
        case let components where components.day! > 1:
            switch format {
            case .concise:
                return NSString(format: NSLocalizedString("DaysAgo", comment: "") as NSString, components.day!) as String
            case .verbose:
                return NSString(format: NSLocalizedString("DaysAgo Verbose", comment: "") as NSString, components.day!) as String
            }
        
        case let components where components.day == 1:
            switch format {
            case .concise:
                return NSLocalizedString("Yesterday", comment: "")
            case .verbose:
                return NSLocalizedString("Yesterday Verbose", comment: "")
            }
        
        case let components where components.hour! > 1:
            switch format {
            case .concise:
                return NSString(format: NSLocalizedString("HoursAgo", comment: "") as NSString, components.hour!) as String
            case .verbose:
                return NSString(format: NSLocalizedString("HoursAgo Verbose", comment: "") as NSString, components.hour!) as String
            }
        
        case let components where components.hour == 1:
            switch format {
            case .concise:
                return NSLocalizedString("HourAgo", comment: "")
            case .verbose:
                return NSLocalizedString("HourAgo Verbose", comment: "")
            }
        
        case let components where components.minute! > 1:
            switch format {
            case .concise:
                return NSString(format: NSLocalizedString("MinutesAgo", comment: "") as NSString, components.minute!) as String
            case .verbose:
                return NSString(format: NSLocalizedString("MinutesAgo Verbose", comment: "") as NSString, components.minute!) as String
            }
        
        case let components where components.minute == 1:
            switch format {
            case .concise:
                return NSLocalizedString("MinuteAgo", comment: "")
            case .verbose:
                return NSLocalizedString("MinuteAgo Verbose", comment: "")
            }
        
        case let components where components.second! > 1 && precision == .seconds:
            switch format {
            case .concise:
                return NSString(format: NSLocalizedString("SecondsAgo", comment: "") as NSString, components.second!) as String
            case .verbose:
                return NSString(format: NSLocalizedString("SecondsAgo Verbose", comment: "") as NSString, components.second!) as String
            }
        
        case let components where components.second == 1 && precision == .seconds:
            switch format {
            case .concise:
                return NSLocalizedString("SecondAgo", comment: "")
            case .verbose:
                return NSLocalizedString("SecondAgo Verbose", comment: "")
            }
        
        default:
            return NSLocalizedString("Now", comment: "")
        }
    }
}
